File: go/local_client/local.go
package main

import (
	"fmt"
	"io"
	"os"
	"strings"
	"time"
)

func main() {
	statbuf, _ := os.Stat("../test_file.txt")
	fmt.Println("File size (in bytes):", statbuf.Size())
	date := statbuf.ModTime().UTC().Format(time.RFC1123)
	fmt.Println("Last modified time:", strings.Replace(date, "UTC", "GMT", 1))

	file, _ := os.OpenFile("../test_file.txt", os.O_RDWR, 0644)
	buf := make([]byte, 5)
	status, _ := file.Read(buf)
	fmt.Println(status, string(buf))
	file.Seek(0, io.SeekStart)
	buf2 := make([]byte, 6)
	status, _ = file.Read(buf2)
	fmt.Println(status, string(buf)+string(buf2))
	file.Write([]byte("word"))
	file.Truncate(30)
	file.Sync()
	file.Close()

	os.Rename("../test_file.txt", "../renamed_file.txt")
	os.Remove("../deleted_file.txt")

	namelist, _ := os.ReadDir("..")
	for _, entry := range namelist {
		if entry.IsDir() {
			fmt.Println(entry.Name() + "/")
		}
		if entry.Type().IsRegular() {
			fmt.Println(entry.Name())
		}
	}
}

File: rust/local_client/local.rs
use std::{
    fs::{OpenOptions, metadata, read_dir, remove_file, rename},
    io::{Read, Seek, SeekFrom::Start, Write},
    str::from_utf8,
    time::SystemTime,
};

use chrono::{TimeZone, Utc};

fn main() {
    let stat = metadata("test_file.txt").unwrap();
    println!("File size (in bytes): {}", stat.len());
    let timestamp = stat
        .modified()
        .unwrap()
        .duration_since(SystemTime::UNIX_EPOCH)
        .unwrap()
        .as_secs();
    let date = Utc
        .timestamp_opt(timestamp as i64, 0)
        .unwrap()
        .to_rfc2822()
        .replace(" +0000", "");
    println!("Last modified time: {date}");

    let mut file = OpenOptions::new()
        .read(true)
        .write(true)
        .open("test_file.txt")
        .unwrap();
    let mut buf = [0u8; 5];
    let status = file.read(&mut buf).unwrap();
    println!("{} {}", status, from_utf8(&buf).unwrap());
    file.seek(Start(0)).unwrap();
    let mut buf2 = [0u8; 6];
    let status = file.read(&mut buf2).unwrap();
    println!(
        "{} {}{}",
        status,
        from_utf8(&buf).unwrap(),
        from_utf8(&buf2).unwrap()
    );
    file.write_all("word".as_bytes()).unwrap();
    file.set_len(30).unwrap();
    file.sync_all().unwrap();

    rename("test_file.txt", "renamed_file.txt").unwrap();
    remove_file("deleted_file.txt").unwrap();

    let namelist = read_dir(".").unwrap();
    for entry in namelist {
        let file = entry.unwrap();
        let filetype = file.file_type().unwrap();
        if filetype.is_dir() {
            println!("{}/", file.file_name().into_string().unwrap());
        }
        if filetype.is_file() {
            println!("{}", file.file_name().into_string().unwrap());
        }
    }
}

// To use Unix domain sockets:
// Server: https://doc.rust-lang.org/std/os/unix/net/struct.UnixListener.html
// Client: https://doc.rust-lang.org/std/os/unix/net/struct.UnixStream.html
